File: skill_sdk/tests/fixtures/js-project/src/shop/payment/PayPalGateway.js
import { PaymentGateway } from './PaymentGateway.js';
import { PaymentResult } from './PaymentResult.js';

/**
 * PayPal-backed payment gateway.
 *
 * @implements {PaymentGateway}
 */
export class PayPalGateway extends PaymentGateway {
  /**
   * @param {string} clientId
   * @param {string} clientSecret
   */
  constructor(clientId, clientSecret) {
    super();
    /** @type {string} */
    this._clientId = clientId;
    /** @type {string} */
    this._clientSecret = clientSecret;
  }

  /**
   * @param {number} amount
   * @param {string} currency
   * @returns {PaymentResult}
   */
  charge(amount, currency) {
    const txId = `pp_ch_${this._hashFloat(amount)}`;
    return new PaymentResult(true, txId, `PayPal charge ${amount} ${currency}`);
  }

  /**
   * @param {string} transactionId
   * @returns {PaymentResult}
   */
  refund(transactionId) {
    return new PaymentResult(
      true,
      `pp_ref_${transactionId}`,
      `PayPal refund ${transactionId}`
    );
  }

  /**
   * @param {number} f
   * @returns {number}
   */
  _hashFloat(f) {
    return Math.floor(f * 100);
  }
}

File: skill_sdk/tests/fixtures/cpp-project/src/shop/PayPalGateway.cpp
#include "PayPalGateway.h"

PayPalGateway::PayPalGateway(const std::string& clientID, const std::string& clientSecret)
    : clientID_(clientID), clientSecret_(clientSecret) {}

PaymentResult PayPalGateway::Charge(double amount, const std::string& currency) {
    return PaymentResult{
        true,
        "pp_ch_" + std::to_string(static_cast<int>(amount * 100)),
        "PayPal charge " + std::to_string(amount) + " " + currency
    };
}

PaymentResult PayPalGateway::Refund(const std::string& transactionID) {
    return PaymentResult{
        true,
        "pp_ref_" + transactionID,
        "PayPal refund " + transactionID
    };
}
